Add EIP: Emit log on revert (#9390)
ethereum/EIPs: EIPS/eip-xxxx.md

---
eip: xxxx
title: Emit log on revert
description: Top level reverts emit a log with revert message
author: Shoham Chakraborty (@shohamc1), Alex Forshtat (@forshtat)
discussions-to: https://ethereum-magicians.org/t/eip-xxxx-emit-log-on-revert/22918
status: Draft
type: Standards Track
category: Core
created: 2025-02-20
---

## Abstract

All calls to the REVERT opcode with non-zero size must emit a log with revert data, making it accessible via standard RPC without the need for tracing.

## Motivation

> Revert messages are currently are inaccessible to users as they are not available via standard RPC. Instead users have to request a node to trace the transaction and check the stack and memory at the moment when the REVERT opcode was executed. 

Reviewer: ```suggestion
Revert messages are currently inaccessible to users as they are not available via standard RPC. Instead, users have to request a node to trace the transaction and check the stack and memory at the moment when the REVERT opcode was executed. 
```

> This introduces overhead for users and nodes - the user must make an additional request to find out why their transaction failed, and the node has to replay the full transaction (which may be slow and computationally expensive) to get back a relatively small piece of data.

Reviewer: ```suggestion
This introduces overhead for users and nodes - users must make an additional request to find out why their transaction failed, and the node has to replay the full transaction (which may be slow and computationally expensive) to get back a relatively small piece of data.
```

Currently it is up to smart wallet developers to emit logs before a revert, however this is not a standard feature and thus cannot be relied upon by tools such as client libraries and block explorers. Making this log part of the protocol allows these tools to rely on logs for revert reasons.

## Specification

### Parameters

* `REVERTTOPIC`: `TBD`
* `DATA_LIMIT`: `TBD`

### Functionality

Whenever `REVERT` is called with non-zero size, emit a log identical to a LOG1 with the topic `REVERTTOPIC`. The log data is the raw bytes of the revert message. The data is truncated to `DATA_LIMIT`.

> ## Rationale
> 
> This is the simples possible implementation that allows revert messages to be accessible via RPC. 

Reviewer: Your sentence introduces the idea well, but it could be strengthened by providing more justification. If the goal is to explain why revert messages should be accessible via RPC, consider expanding on why this implementation is preferable. For example, highlighting its efficiency, ease of integration, or benefits for developers would make the rationale more compelling.

```suggestion
## Rationale

This is the simplest possible implementation that allows revert messages to be accessible via RPC. 
```

Author: Added a bit more to the rationale, let me know if you think if it needs more details.

## Backwards Compatibility

Reverted transactions may now contain up to one log.

## Security Considerations

> Reverted transactions must cost atleast intrinsic gas which is atleast 21000 gas, which is much more expensive than the LOG1 opcode (750 gas). Hence, this EIP does not introduce any new avenues to inflate Ethereum storage requirements. However, it is expected to increase the average number of logs.

Reviewer 2: ```suggestion
Reverted transactions must cost at least intrinsic gas (21000 gas), which is much more expensive than the LOG1 opcode (750 gas). Hence, this EIP does not introduce any new avenues to inflate Ethereum storage requirements. However, it is expected to increase the average number of logs.
```
